tree/tree.go

```diff
@@ -0,0 +1,158 @@
+	t  *arbo.Tree
+	db db.Database
+}
+
+// Options is used to pass the parameters to load a new Tree
+type Options struct {
+	// Path of the db.Database to be used
+	Path string
+	// MaxLevels that will have the Tree
+	MaxLevels int
+	// HashFunc defines the hash function that the tree will use
+	HashFunc arbo.HashFunction
+}
+
+// New returns a new Tree, if there already is a Tree in the database, it will
+// load it
+func New(opts Options) (*Tree, error) {
+	database, err := badgerdb.New(badgerdb.Options{Path: opts.Path})
+	if err != nil {
+		return nil, err
+	}
+
+	tree, err := arbo.NewTree(database, opts.MaxLevels, opts.HashFunc)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Tree{
+		t:  tree,
+		db: database,
+	}, nil
+}
+
+func (t *Tree) DB() db.Database {
+	return t.db
+}
+
+// Get returns the value for a given key.
+func (t *Tree) Get(rTx db.ReadTx, key []byte) ([]byte, error) {
+	_, v, err := t.t.GetWithTx(rTx, key)
+	if err != nil {
+		return nil, err
+	}
+
+	return v, nil
+}
+
+// Set adds or updates a key. First performs a Get, and if the key does not
+// exist yet it will call Add, while if the key already exists, it will perform
+// an Update. If the non-existence or existence of the key is already known, is
+// more efficient to directly call Add or Update.
+func (t *Tree) Set(wTx db.WriteTx, key, value []byte) error {
+	_, err := t.Get(wTx, key)
+	if err == arbo.ErrKeyNotFound {
+		// key does not exist, use Add
+		return t.Add(wTx, key, value)
+	} else if err != nil {
+		return err
+	}
+	// key already exists, use Update
+	return t.t.UpdateWithTx(wTx, key, value)
+}
+
+// Add adds a new leaf, if the key already exists will return error
+func (t *Tree) Add(wTx db.WriteTx, key, value []byte) error {
+	return t.t.AddWithTx(wTx, key, value)
+}
+
+// AddBatch adds a batch of key-values to the Tree. Returns an array containing
+// the indexes of the keys failed to add. Supports empty values as input
+// parameters, which is equivalent to 0 valued byte array.
+func (t *Tree) AddBatch(wTx db.WriteTx, keys, values [][]byte) ([]int, error) {
+	return t.t.AddBatchWithTx(wTx, keys, values)
+}
+
+func (t *Tree) Iterate(callback func(key, value []byte) bool) error {
+	callbackWrapper := func(currLvl int, k, v []byte) bool {
+		callback(k, v)
+		return true
+	}
+	rTx := t.db.ReadTx()
+	root, err := t.Root(rTx)
+	if err != nil {
+		return err
+	}
+	return t.t.IterateWithStop(root, callbackWrapper)
+}
+
+// Root returns the current root.
+func (t *Tree) Root(rTx db.ReadTx) ([]byte, error) {
+	return t.t.RootWithTx(rTx)
+}
+
+// Size returns the number of leafs under the current root
+func (t *Tree) Size(rTx db.ReadTx) uint64 {
+	return 0
+}
+
+// GenProof returns a byte array with the necessary data to verify that the
+// key&value are in a leaf under the current root
+func (t *Tree) GenProof(rTx db.ReadTx, key []byte) ([]byte, error) {
+	// TODO NOTE leafK & leafV will be needed for verifying the proof,
+	// maybe is worth to return them
```

Comment:
@ed255 check note

Reply:
I understand that right now this function returns only the siblings serialized right?  And that it only returns proofs of existence.

Is `leafK == key`? (Where `key` is in the function arguments).  If so, I don't think we need to return the key, because the user that generates the proof already has the key.

About the `leafV` I think it's worth returning it, as it's needed for verification.

And in any case, I think these extra values should be in different return values (that is, not inside the `[]byte` proof).

Reply:
`leafK` is not the same than the `key` passed as input in the cases where the key does not exist, but the path ends in another leaf, in that case. This is the case to generate a proof of non-existence. So for these cases, the user does not know the key that will be needed to verify the proof before hand.

I'll add to the return the `leafL` and `leafV`.